File: test/analytics/analytics_contract_test.dart
import 'package:exam_coach/analytics/analytics_events.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('accepts registered events with allow-listed scalar properties', () {
    final properties =
        AnalyticsEvents.validate(AnalyticsEvents.questionAnswered, {
          'sessionId': 'session_1',
          'questionId': 'question_1',
          'taxonomyNodeId': 'topic_1',
          'isCorrect': true,
          'timeSpentMs': 1200,
        });

    expect(properties['isCorrect'], isTrue);
  });

  test('rejects unknown events, properties, and oversized strings', () {
    expect(
      () => AnalyticsEvents.validate('unknown_event', const {}),
      throwsArgumentError,
    );
    expect(
      () => AnalyticsEvents.validate(AnalyticsEvents.resultViewed, const {
        'email': 'private@example.com',
      }),
      throwsArgumentError,
    );
    expect(
      () => AnalyticsEvents.validate(AnalyticsEvents.resultViewed, {
        'sessionId': 'x' * 241,
      }),
      throwsArgumentError,
    );
  });
}

File: functions/test/analytics_contract.test.js
"use strict";

const assert = require("node:assert/strict");
const test = require("node:test");
const {validateAnalyticsProperties} = require("../analytics_contract");

test("accepts allow-listed scalar analytics properties", () => {
  assert.deepEqual(
    validateAnalyticsProperties("result_viewed", {sessionId: "session_1"}),
    {sessionId: "session_1"},
  );
});

test("rejects unknown or privacy-risk analytics data", () => {
  assert.throws(
    () => validateAnalyticsProperties("unknown_event", {}),
    /not registered/,
  );
  assert.throws(
    () => validateAnalyticsProperties("result_viewed", {email: "a@example.com"}),
    /not allowed/,
  );
  assert.throws(
    () => validateAnalyticsProperties("result_viewed", {sessionId: "x".repeat(241)}),
    /unsafe/,
  );
});
